Snap Support (etcd 2.x/3.x forward facing)

lazypower opened the pull request:
This introduces a large refactoring of the layer codebase to be compliant with Snaps and the data-paths for Snap packages. The charm strives to use strict confinement primitives and enable the snaps to do complex operations at scale using only template logic and minimal intervention.

Notable Updates in this branch:
- Introduces a manual path for snap-upgrade, to ensure the operator is aware of this change
- Migration path assumes the snap-store, or manual attachment of snaps in offline environment(s)
- Refactors external data mounts to reside in /media per the 'removable-media' interface in snaps.
- Supports both etcd 3.x and 2.x
- refactors critical-path actions to python from bash

Note: in order for this branch to be properly used, the etcd-snap will require  https://github.com/markshuttle/etcd-snaps/pull/4  before any external storage can be used. Snap confinement will block the process otherwise.

Commit:
WIP - Working multi-snap cluster

Initial deployment (non-upgrade) code. Functional X node cluster. Does not survive leader termination.

Commit:
Adds migration process from apt to snap

Commit:
Actions snap refactor

Commit:
Tweaked peer health string

Commit:
Adds etcd3 snap support (prelim)

Tested with 3.1.2.

- Refactors template rendering to the version listed as installed in
snapd.
- Adds etcd2/3 configuration templates
- Readme updates
- Removing amulet package from test dependencies.

Commit:
Cleanup of restore and snap-upgrade work

- Move "migrate.py" work to snap-upgrade
- Refactores restore to be python based and works with both 2.x and 3.x
series.
- Changes default casing in actions.yaml bool params.

Commit:
Refactor storage to work with $SNAP_DATA

Storage was historically temp mounted, migrated, then blindly mounted
over the default $ETCD_DATA path. This now refactores the storage to
work with snaps that have access to the 'pluggable-media' interface.

This particular feature depends on
https://github.com/markshuttle/etcd-snaps/pull/4

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+
+def start_etcd_forked():
+    ''' Start the etcd daemon temporarily to initiate new cluster details '''
+    raw = "/snap/etcd/current/bin/etcd -data-dir={0} -force-new-cluster"
```
Does etcd recursively create the data-dir if it doesn't exist?

lazypower replied:
Not necessarily. As its working with snap confinement, it can and will be blocked if its configured outside of /media or $SNAP_DATA/$SNAP_COMMON/$SNAP_USER

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+
+def pkill_etcd():
+    ''' Kill the temporary forked etcd daemon '''
+    cmd = 'pkill etcd'
```
Can we kill by PID instead? Little nitpicky, but feels better. I think the Popen object includes the pid.

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+    ''' Block until the forked etcd instance has started and return'''
+    output = b""
+    loop = 0
+    MAX_WAIT = 10
```
Do you think this should be configurable?

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+
+if __name__ == '__main__':
+    preflight_check()
+    stop_etcd()
```
It doesn't look like we're bailing here if we fail to stop etcd. Do we need to?

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+        log('Failed to stop service: etcd')
+
+
+def shasum_file(filepath):
```
Really nicely done. No memory issues, eh? :+1:

mbruzek replied:
Good use of the hashlib here. @wwwtyro 's comment makes me think about shelling out to sha256sum in this case due to memory concerns. But you do whichever you want.

wwwtyro replied:
Ah, my bad. I was trying to communicate that this intelligently takes memory into account, and so shouldn't have a problem. Clearly, I failed!

mbruzek replied:
If the etcd data > total RAM I would worry about memory. I just don't know how large the etcd data is.

wwwtyro replied:
@mbruzek Yeah exactly! That's why I like this, he's chunking it into 64KB blocks. :)

mbruzek replied:
My understanding of the `update()` method is wrong, I withdraw my concern and comment. Great job here Chuck.

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+    with chdir(ETCD_DATA_DIR):
+        if not SKIP_BACKUP:
+            log('Backing up existing data found in {}'.format(ETCD_DATA_DIR))
+            # tar cvf $TARGET_PATH/$ARCHIVE default
```
Does this need to happen?

lazypower replied:
Oh man I'm so happy we have code review. Whoops!

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+def reconfigure_client_advertise():
+    ''' Reconfigure the backup to use host network addresses for client advertise
+        instead of the assumed localhost addressing '''
+    cmd = "/snap/bin/etcd.etcdctl member list"
```
Nitpicky, but you use this string twice (other instance on line 87). No objection if you want to define a variable for it.

wwwtyro commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+
+def render_backup():
+    ''' Backup existing data in the event of restoration on a dirty unit. '''
+    if not os.path.isdir(ETCD_DATA_DIR):
```
Since we're setting SKIP_BACKUP to default true, should we log if there's no data dir and SKIP_BACKUP is false?

wwwtyro commented on actions/snap-upgrade.py:
```diff
@@ -0,0 +1,144 @@
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # Create a temporary path to perform the backup, and date the contents.
+        dated_path = "{0}/etcd_migration_{1}".format(tmpdir, datestring)
+        os.makedirs(dated_path, exist_ok=True)
```
Is it really okay if this already exists? Seems like it would be a weird case?

wwwtyro commented on actions/snap-upgrade.py:
```diff
@@ -0,0 +1,144 @@
+    check_call(split(cmd))
+    for f in deb_paths['config']:
+        log('Removing file {}'.format(f), 'INFO')
+        os.remove(f)
```
I think this throws an error if the file doesn't exist. Do you want to deal with that here?

mbruzek commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+TARGET_PATH = action_get('target')
+
+
+def preflight_check():
```
Since we are going to snaps do you need to add checks if you can write in the "etcd_data_dir" or something of that nature? I see that is a layer option and if someone did not know about snaps or confinement they may try to configure the directory incorrectly.

marcoceppi replied:
I suppose so, but since layer options are a compile time flag and not a runtime normal users aren't exposed to it.

mbruzek commented on actions/restore.py:
```diff
@@ -0,0 +1,147 @@
+
+
+if __name__ == '__main__':
+    preflight_check()
```
This is awesome code! Very clear steps that are easy to follow. Wondering if it is worth it printing each step out so the user knows what is going or how each step went.

mbruzek commented on reactive/etcd.py:
```diff
@@ -155,30 +140,28 @@ def leader_config_changed():
         # Render just the leaders configuration with the new values.
-        render('defaults', '/etc/default/etcd', bag.__dict__, owner='root',
-               group='root')
+        render_config()
```
You are rendering configuration here and _not_ passing the bag variable. Did you mean to do that or not? I don't see bag getting any values assigned in this method but the other uses appear to take the bag as a parameter.

lazypower replied:
Passing in Bag is only really necessary when you have changes to the bag object (like updating a cluster string with a seeded value from leader_data, as seen in line: 315. 

Good call out, but I think this is an acceptable time to use the objects probed defaults.

mbruzek reviewed (approved):
I looked over this code it looks very clean and clear! Really liked how the code reads, well done sir! None of my comments or questions are blockers. 

+1 from me. Looking forward to using the new etcd.

Commit:
Cleanup of Unit Tests

Changes were introduced in the snap refactoring that change the
executable name. The tests needed to be updated to reflect this change.

marcoceppi reviewed (changes requested):
Great start, I think some of the feedback needs to be addressed

Commit:
Review feedback from @wwtyro and @mbruzek

I've addressed the review callouts on error handling in the actions.
Additionally I've bumped the test to default to testing from the
3.0/stable channel of the snap pending 3.1.2 landing in the snap
channel for 3.1/stable

Commit:
Add missing install action for forced snap upgrade paths when using
resources.

Commit:
Add snap migration readme instructions

Commit:
Correct the health action to use the snap paths and non-tls wrapped
localhost binding

Commit:
Readme bump w/ notes for resource requirements during the restore action

Commit:
Snap migration updates

- Updates readme with caveats
- Corrects some bugs found in the snap-upgrade action
- Corrected the placement of tls keys post upgrade

Commit:
Adjusted the permissions on client TLS keys

Commit:
Change from using server certs (with client auth) to client
certificates.

Commit:
Refactor chmod bits for ubuntu user

Commit:
My testing gave me some ideas for the README.

Commit:
Merge pull request #2 from mbruzek/snap-readme-update

My testing gave me some ideas for the README.
